### PR TITLE
Make slider path parsing more friendly with spans

### PR DESCRIPTION
**Pull request by huoyaoyuan**
In order to unblock #23282, this PR refactors slider parsing to not passing substrings around.

The majority of existing logic is preserved. The performance shouldn't change much with this PR, just ensuring there's no regression.

Before:

| Method                  | Mean     | Error     | StdDev    | Gen0     | Gen1     | Allocated |
|------------------------ |---------:|----------:|----------:|---------:|---------:|----------:|
| BenchmarkBundledBeatmap | 1.529 ms | 0.0083 ms | 0.0065 ms | 228.5156 | 187.5000 |   4.13 MB |

After:

 | Method                  | Mean     | Error     | StdDev    | Gen0     | Gen1     | Allocated |
|------------------------ |---------:|----------:|----------:|---------:|---------:|----------:|
| BenchmarkBundledBeatmap | 1.540 ms | 0.0208 ms | 0.0195 ms | 228.5156 | 187.5000 |   4.12 MB |

### CODE REVIEW

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs [hunk 1]
```diff
@@ -266,68 +267,81 @@ private PathControlPoint[] convertPathString(string pointString, Vector2 offset)
...
-                string endPoint = endIndex < pointSplit.Length - 1 ? pointSplit[endIndex + 1] : null;
+                var controlPoints = new List<ArraySegment<PathControlPoint>>(pointsCount);
+
+                for (int i = 0; i < segmentsCount; i++)
+                {
+                    int startIndex = segments[i].StartIndex;
+                    int endIndex = i < segmentsCount - 1 ? segments[i + 1].StartIndex : pointsCount;
+                    Vector2? endPoint = i < segmentsCount - 1 ? points[endIndex] : null;
+                    controlPoints.AddRange(convertPoints(segments[i].Type, new ArraySegment<Vector2>(points, startIndex, endIndex - startIndex), endPoint));
+                }
```
[the remark this answers is not in the document]
Comment (huoyaoyuan) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:303]: I initially considered folding the `convertPoints` method. Now I don't want to make further change aside from unblock string parsing, it is better to separate the two conditions.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs [hunk 2]
```diff
@@ -266,68 +267,81 @@ private PathControlPoint[] convertPathString(string pointString, Vector2 offset)
...
-                vertices[i] = new PathControlPoint();
+            var vertices = new PathControlPoint[points.Count];
 
             // Parse into control points.
-            for (int i = 1; i < points.Length; i++)
-                readPoint(points.Span[i], offset, out vertices[readOffset + i - 1]);
-
-            // If an endpoint is given, add it to the end.
-            if (endPoint != null)
-                readPoint(endPoint, offset, out vertices[^1]);
```
[the remark this answers is not in the document]
Comment (huoyaoyuan) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:323]: Correct. It took some time for me to figure out what `endPoint` is for.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs [hunk 3]
```diff
@@ -266,68 +267,81 @@ private PathControlPoint[] convertPathString(string pointString, Vector2 offset)
...
             string[] pointSplit = pointString.Split('|');
 
-            var controlPoints = new List<Memory<PathControlPoint>>();
-            int startIndex = 0;
-            int endIndex = 0;
-            bool first = true;
+            var points = ArrayPool<Vector2>.Shared.Rent(pointSplit.Length);
+            var segments = ArrayPool<(PathType Type, int StartIndex)>.Shared.Rent(pointSplit.Length);
+            int pointsCount = 0;
+            int segmentsCount = 0;
```
[the remark this answers is not in the document]
Comment (huoyaoyuan) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:273]: Honestly, the `buffer[count++]` pattern looks very common to me and I didn't think it's hard to read.

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs]: Combined with the rest of the logic and the various kind of collections that exist in the logic (points string array, points array, segments array), it becomes very cryptic for me.

**Review by frenzibyte — APPROVED** [review]
Looks good to both my eyes and CI.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs [hunk 4]
```diff
@@ -264,70 +265,93 @@ private PathType convertPathType(string input)
...
-                controlPoints.AddRange(convertPoints(pointSplit.AsMemory().Slice(startIndex, endIndex - startIndex), endPoint, first, offset));
-                startIndex = endIndex;
-                first = false;
+                for (int i = 0; i < segmentsCount; i++)
+                {
+                    if (i < segmentsCount - 1)
+                    {
+                        int startIndex = segmentsBuffer[i].StartIndex;
+                        int endIndex = segmentsBuffer[i + 1].StartIndex;
+                        controlPoints.AddRange(convertPoints(segmentsBuffer[i].Type, allPoints.Slice(startIndex, endIndex - startIndex), pointsBuffer[endIndex]));
```
Comment (huoyaoyuan) [inline osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:306]: Specifically, I think *this* particular case is better readable with range expression.

I don't have preference on other cases though.

Comment (frenzibyte) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:306]: There have been multiple occurrences of the `..` operator being used and us not liking it in review, I don't want that happening here so I changed it with the `Slice` method to move the PR forward.

**Comment by frenzibyte** [comment]
Requesting a second pair of eyes before proceeding with a merge, the PR is good but I'm not sure if there's something I should've checked against before approval (like triggering a diffcalc command to make sure nothing has changed with beatmap parsing or something).

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by frenzibyte** [comment]
Spreadsheets show no difference as well, so let's get this in.

**Pull request merged**